### ISSUE TITLE
vSphere Cloud Provider not working

### ISSUE THREAD

**Issue by MaxDiOrio** [rancher/rancher#14889]
**Rancher versions:**
rancher/server or rancher/rancher:  v2.0.6

**Docker version: (`docker version`,`docker info` preferred)**  Docker version 1.12.6, build 3e8e77d/1.12.6 

**Operating system and kernel: (`cat /etc/os-release`, `uname -r` preferred)**.   RHEL 7.5 3.10.0-862.9.1.el7.x86_64

**Type/provider of hosts: (VirtualBox/Bare-metal/AWS/GCE/DO)**. VMWare vSphere

**Setup details: (single node rancher vs. HA rancher, internal DB vs. external DB)**. Single Node

So I'm trying to deploy clusters with vSphere Cloud Provider so that I can create persistent volumes.

When I bring up the cluster after editing the yaml, all machines are stuck Registering - Waiting to register with Kubernetes

I see the following in the Rancher logs.  The really strange one is why is it looking for the datacenter as a uint?  

2018/08/01 19:32:33 [ERROR] cluster [c-hlnt5] provisioning: [workerPlane] Failed to bring up Worker Plane: Failed to verify healthcheck: Failed to check https://localhost:10250/healthz for service [kubelet] on host [10.85.175.34]: Get https://localhost:10250/healthz: dial tcp [::1]:10250: getsockopt: connection refused, log: F0801 19:32:20.783951    7019 server.go:233] failed to run Kubelet: could not init cloud provider "vsphere": failed to parse "Latham-Colo-DC" as uint: expected integer

[workerPlane] Failed to bring up Worker Plane: Failed to verify healthcheck: Failed to check https://localhost:10250/healthz for service [kubelet] on host [10.85.175.34]: Get https://localhost:10250/healthz: dial tcp [::1]:10250: getsockopt: connection refused, log: + umount /var/lib/docker/overlay/92e64d20058ccfab51ab2f4e374b797e330d9c5b09761f33a7b4b51e849864e3/merged/opt/rke/var/lib/kubelet

Failed to create Certificates deployer container on host [10.85.175.36]: Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?

I'm using the following Cloud Provider yaml:

```
cloud_provider:
  name: vsphere
  vsphereCloudProvider:
      global: 
        insecure-flag: 1
        user: svcaccount
        password: password
        port: 443
      virtual_center:
        10.85.128.181:
            datacenters: Latham-Colo-DC
      workspace:
        server: 10.85.128.181
        datacenter: Latham-Colo-DC
        folder: pvols
        default-datastore: /Latham-Colo-DC/datastore/vmware_kube1
        resourcepool-path: /Latham-Colo-DC/host/Production-Cluster
      disk:
        scsicontrollertype: pvscsi
```

Any ideas?  Thanks!

**Comment by MaxDiOrio** [comment]
I tracked this one down after looking at a separate K8 cluster I brought up by hand.  You have a typo in your documentation:

https://rancher.com/docs/rke/v0.1.x/en/config-options/cloud-providers/vsphere/

The key under virtual_center should be **datacenter**, not datacenter**s**

As soon as I changed it and deployed a new cluster, it started working perfectly.

**Comment by edqallen** [comment]
@MaxDiOrio can you update/respond to indicate what k8s documentation you used to create your YAML file?  I'm in a similar boat here, except my vmware enterprise cluster is using SDRS on a storage cluster and I'm having trouble getting the options correct.

**Comment by B3DTech** [comment]
I'm ooo until Thursday, but as long as you follow the Rancher doc I linked to, including setting up the service account it works fine. Just drop the S from the datacenters like I mentioned. I still am unable to create persistent disks however. They never get created.

**Comment by lynic** [comment]
I notice that if choosing vsphere, then there's no way to set cloud provider to vsphere. The only two options are None and custom, and after choosing custom, there's no way to input parameters.

**Comment by MaxDiOrio** [comment]
That's correct. They changed it to manually editing cloud provider settings in the yaml only.

Sent from Nine<http://www.9folders.com/>
________________________________

**Comment by lynic** [comment]
After reading #13115 , now I understand how to setup it correctly.

**Comment by superseb** [comment]
According to https://github.com/kubernetes/kubernetes/blob/release-1.10/pkg/cloudprovider/providers/vsphere/vsphere.go#L90, the key should be `datacenters`. I think you are hitting https://github.com/rancher/rke/commit/6b7ab838992b1202f26285a534f4b797cbc3fb70#diff-e2a140aff74f8d1d81845bb3e858ae95 which is fixed in v2.0.7. Can you try with that version and let us know if it doesn't work?
